Introduce auto cleanup (administration domain) #19

de-jcup opened the pull request:
This PR
- closes #19
- closes #1059
- closes #1072 
- closes #1073
- closes #1076
- closes #1077
- closes #1124

Jeeppler commented on sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/log/ProjectScanLog.java:
```diff
@@ -58,6 +58,8 @@
     public static final String STATUS_FAILED = "FAILED";
     public static final String STATUS_ABANDONED = "ABANDONDED";
 
+    public static final String QUERY_DELETE_LOGS_OLDER_THAN = "DELETE FROM ProjectScanLog l WHERE l." + PROPERTY_STARTED + " <:cleanTimeStamp";;
```
```suggestion
    public static final String QUERY_DELETE_LOGS_OLDER_THAN = "DELETE FROM ProjectScanLog l WHERE l." + PROPERTY_STARTED + " < :cleanTimeStamp";
```

Furthermore, the difference between the letter `l` and `1` is difficult to identify. To go around this issue, I would prefer `log`. It would then be `ProjectScanLog log WHERE log`…

Jeeppler commented on sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/report/ScanReport.java:
```diff
@@ -43,6 +43,9 @@
     public static final String CLASS_NAME = ScanReport.class.getSimpleName();
+    public static final String PROPERTY_REPORT_STARTED = "started";
+
+    public static final String QUERY_DELETE_REPORTS_OLDER_THAN = "DELETE FROM ScanReport r WHERE r." + PROPERTY_REPORT_STARTED + " <:cleanTimeStamp";
```
```suggestion
    public static final String QUERY_DELETE_REPORTS_OLDER_THAN = "DELETE FROM ScanReport r WHERE r." + PROPERTY_REPORT_STARTED + " < :cleanTimeStamp";
```

Jeeppler commented on sechub-scan/src/test/java/com/mercedesbenz/sechub/domain/scan/ScanMessageHandlerTest.java:
```diff
@@ -32,32 +34,50 @@
 
     @Test
-    public void when_sending_message_id_PROJECT_ACCESS_LEVEL_CHANGED_changeProjectAccessLevel_is_called() {
+    void handler_receiving_auto_cleanup_calls_config_serice_with_message_data() {
```
```suggestion
    void handler_receiving_auto_cleanup_calls_config_service_with_message_data() {
```

Jeeppler reviewed (changes requested):
First part of the review.

I was unable to find tests which test negative values for the deletion time. I would like to see a test, testing for `-1` days or months to make sure negative values are not accepted.

Jeeppler commented on sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/messaging/AdministrationConfigMessage.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ */
+@JsonIgnoreProperties(ignoreUnknown = true) // we do ignore to avoid problems from wrong configured values!
+@MustBeKeptStable("This configuration is used by communication between (api) scan,report, scheduler and administration")
```
```suggestion
@MustBeKeptStable("This configuration is used by communication between (api) scan, report, scheduler and administration")
```

Jeeppler commented on sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/usecases/admin/config/UseCaseAdminFetchesAutoCleanupConfiguration.java:
```diff
@@ -0,0 +1,27 @@
+		id=UseCaseIdentifier.UC_ADMIN_FETCHES_AUTO_CLEANUP_CONFIGURATION,
+		group=UseCaseGroup.CONFIGURATION,
+		apiName="adminFetchesAutoCleanupConfiguration",
+		title="Admin fetchesauto cleanup configuration",
```
```suggestion
		title="Admin fetches auto cleanup configuration",
```

Jeeppler commented on sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/usecases/autocleanup/UseCaseAdministrationAutoCleanExecution.java:
```diff
@@ -0,0 +1,28 @@
+		group=UseCaseGroup.TECHNICAL,
+		apiName=APIConstants.NO_API_AVAILABLE,
+		title="Sechub administration domain auto cleanup",
+		description="The administration domain does execute an auto cleanup. Periodically it will be checked if there is any outdated job information. The timeperiod is detrmine")
```
```suggestion
		description="The administration domain does execute an auto cleanup. Periodically it will be checked if there is any outdated job information. The time period is determine")
```

Jeeppler commented on sechub-shared-kernel/src/main/java/com/mercedesbenz/sechub/sharedkernel/usecases/autocleanup/UseCaseScanAutoCleanExecution.java:
```diff
@@ -0,0 +1,28 @@
+		group=UseCaseGroup.TECHNICAL,
+		apiName=APIConstants.NO_API_AVAILABLE,
+		title="Sechub scan domain auto cleanup",
+        description="The scan domain does auto cleanup old data.This is done periodically.")
```
```suggestion
        description="The scan domain does auto cleanup old data. This is done periodically.")
```

Jeeppler commented on sechub-shared-kernel/src/test/java/com/mercedesbenz/sechub/sharedkernel/TimeCalculationServiceTest.java:
```diff
@@ -0,0 +1,45 @@
+    @NullSource
+    @CsvSource({ "0", "1", "90", "365" })
+    @ParameterizedTest
+    void now_minus_days_calulation_works(Long days) {
```
What happens if one deducts so many days, that the result will be negative? Do I get an exception?

de-jcup replied:
The time calculation service does not know that negative values are not acceptable. it will add them.

I added tests to ensure the behaviour.

I changed the implementation inside configuration update mechanism and also inside the execution + added tests for this.
So
- it cannot be configured with negative values by admins
- even when somebody changes value in database, the execution will just log an error but not accept it any longer.

Thanx for finding that.

Jeeppler commented on sechub-testframework/src/main/java/com/mercedesbenz/sechub/test/TestConstants.java:
```diff
@@ -0,0 +1,13 @@
+
+    public static final String DESCRIPTION_DISABLED_BECAUSE_A_MANUAL_TEST_FOR_GENERATION = "This test can be used by developers to generate content. "
+            + "Not necesary for normal CI tests. A developer can run the 'test' to produce very fast generated outtput, "
+            + "without starting the spring boot application or any gradle task(so much faster)";
```
```suggestion
            + "without starting the SpringBoot application or any Gradle task (so much faster)";
```

de-jcup replied:
I changed to "Spring Boot" and "Gradle" 

but our suggestion "SpringBoot" is not correct. They write it not this way: see https://spring.io/projects/spring-boot

Jeeppler reviewed (changes requested):
See comments.

de-jcup commented:
> I was unable to find tests which test negative values for the deletion time. I would like to see a test, testing for `-1` days or months to make sure negative values are not accepted.

Thank you, this was not only missing tests, but an implementation failure.

I added 
- `TimeCalculationServiceTest#now_minus_days_calulation_works` 
  (the calculation service WILL ADD in case of negative values, it's responsible for general calculation and shall be reusable)
- `AdministrationAutoCleanupServiceTest#cleanup_executes_NOT_delete_job_information_for_minus_1_day`
  (also changed implementation, because there was no check)
- and some tests in `AdministrationConfigServcieTest` also about minus days - those configurations may not be accepted
- I also changed implementation parts in update service but also in execution time, so problem can not happen any longer

de-jcup commented on sechub-doc/src/main/java/com/mercedesbenz/sechub/docgen/reflections/Reflections.java:
```diff
@@ -211,7 +212,7 @@ void fetchJavaClassNamesToInspect(File sourceDirectory, File file, JavaContentFi
                 classesToInspect.add(clazz);
             } catch (ClassNotFoundException e) {
-                LOG.warn("Ignored class:{}", className);
+                LOG.trace("Class not found in sechub-doc classpath, only on source code:{}", className);
```
I added a comment inside code + changed the log output, so it is more clear what is meant. Please look inside - I think it is now more clear.

de-jcup replied:
I also changed back from trace to warn - maybe we should (inside another issue) - change the classpath of `sechub-doc`.

```
10:26:55.715 [main] WARN com.mercedesbenz.sechub.docgen.reflections.Reflections - The class 'com.mercedesbenz.sechub.analyzer.cli.SecHubAnalyzerApplication' was not found in ...
```
Some of the warnings can be ignored, but when we add some new sub projects and the documentation does not contain it, we should have at least a log output for debugging...

de-jcup commented on sechub-doc/src/test/java/com/mercedesbenz/sechub/docgen/messaging/MessagingAnnotationsCorrectDefinedTest.java:
```diff
@@ -0,0 +1,338 @@
+        /* test */
+        for (MessageID message : MessageID.values()) {
+            if (!allSentMessages.contains(message)) {
+                if (context.messagesAnswerResultSynchronous.contains(message)) {
```
It was not clear for myself if the changes you proposed would lead to exact same test - but nevertheless I think the test is not very readable (even for myself now).

So I decided to change the test and make it more readable - and remove also the `continue`. Please look at the changes.

de-jcup commented on sechub-integrationtest/src/main/java/com/mercedesbenz/sechub/integrationtest/api/TestAPI.java:
```diff
@@ -997,4 +1010,93 @@ public static void waitProjectDoesNotExist(TestProject project) {
+
+    /**
+     * Will ensure auto cleanup configuration is set to given days and that every
+     * domain is synched.
```
I prefer synched. Will keep it as is.

de-jcup commented on sechub-integrationtest/src/main/java/com/mercedesbenz/sechub/integrationtest/api/TestAPI.java:
```diff
@@ -997,4 +1010,93 @@ public static void waitProjectDoesNotExist(TestProject project) {
+
+            waitUntilAutoCleanupConfigurationChangedTo(data);
+        }
+        waitUntilEveryDomainHasAutoCleanupSynchedToDays(days);
```
Will keep ch

de-jcup commented on sechub-integrationtest/src/test/java/com/mercedesbenz/sechub/integrationtest/scenario1/AutoCleanupScenario1IntTest.java:
```diff
@@ -0,0 +1,94 @@
+
+        /* test */
+        /* @formatter:off */
+        waitUntilEveryDomainHasAutoCleanupSynchedToDays(1);
```
will keep

de-jcup commented on sechub-integrationtest/src/test/java/com/mercedesbenz/sechub/integrationtest/scenario1/AutoCleanupScenario1IntTest.java:
```diff
@@ -0,0 +1,94 @@
+    public void auto_cleanup_NEVER_executed_in_any_domain_when_admin_configures_cleanupdays_0() {
+
+        /* prepare */
+        resetAutoCleanupDays(4711);// so different to 0 which we will set next
```
We have spotless and I rely on spotless. 
Such things like a space between anything for code not being inside "formatter off" region will not be fixed by me any longer.
So I will NOT do this.

If is possible to change the eclipse formatter settings this way and to provide this in spotless via the formatter xml - fine - lets do it, but I will not do any such manual formatting changes.

Jeeppler commented on sechub-schedule/src/main/java/com/mercedesbenz/sechub/domain/schedule/SchedulerJobBatchTriggerService.java:
```diff
@@ -90,9 +90,10 @@ protected void postConstruct() {
-            + "time shifted, simply change the initial delay values in your wanted way.")
+    @MustBeDocumented(value = "Job scheduling is triggered by a cron job operation - default is 10 seconds to delay after last execution. "
+            + "The initial delay is " + DEFAULT_INITIAL_DELAY_MILLIS
+            + " milliseconds is defined. It can be configured different,so when you need to startup a cluster "
```
The second part is difficult to understand. How about:

It can be configured differently. This is useful when you need to startup a cluster. Simply change the initial delay values in to allow the cluster to startup.

de-jcup replied:
You are right about the last sentence.
- I also fixed here the grammar issue like done before for `AutoCleanupConstants#TRIGGER_STEP_MUST_BE_DOCUMENTED` .
- I also fixed the last sentence in `AutoCleanupConstants#TRIGGER_STEP_MUST_BE_DOCUMENTED` .

Jeeppler reviewed (changes requested):
See comments.

de-jcup commented on sechub-doc/src/main/java/com/mercedesbenz/sechub/docgen/spring/ScheduleDescriptionGenerator.java:
```diff
@@ -57,7 +57,7 @@ public String generate(ClasspathDataCollector collector) {
             SortedSet<TableRow> table = entries.getValue();
 
-            sb.append("[options=\"header\",cols=\"1,1,1,1\"]\n");
+            sb.append("[options=\"header\",cols=\"1,1,1\"]\n");
```
When you do not align columns they are always same size: 

![image](https://user-images.githubusercontent.com/11644753/157815969-4d2f4f0a-ed69-4fac-9d0d-7eaf3fbf6df0.png)

Currently we do more or less the same when using "cols=1,1,1":
![image](https://user-images.githubusercontent.com/11644753/157818277-f36a4255-cd09-47f2-bbd0-81f17b252f3a.png)

If we would keep this, we could really remove the cols definition - your are right.

But I think wasting here so much space is not good here: "Type" is always only some small text but column has same size and the others need more .
So I have changed the cols attribute now to "cols=1,6,6" and it looks much better (and it makes sense to use cols here):
![image](https://user-images.githubusercontent.com/11644753/157821081-2d34ee23-973b-4dbb-9022-9dcd36118ead.png)

Jeeppler reviewed (approved):
Looks good to me.
